orchagent/dash/dashrouteorch.cpp

```diff
@@ -0,0 +1,914 @@
+    {
+        return true;
+    }
+
+    qos_entries_[qos_name] = entry;
+    SWSS_LOG_INFO("Added QOS entries for %s", qos_name.c_str());
+
+    return true;
+}
+
+bool DashRouteOrch::removeQosEntry(string qos_name)
+{
+    SWSS_LOG_ENTER();
+
+    if (qos_entries_.find(qos_name) == qos_entries_.end())
+    {
+        return true;
+    }
+
+    qos_entries_.erase(qos_name);
+    SWSS_LOG_INFO("Removed QOS entries for %s", qos_name.c_str());
+
+    return true;
+}
+
+void DashRouteOrch::doTaskQosTable(Consumer& consumer)
+{
+    auto it = consumer.m_toSync.begin();
+    while (it != consumer.m_toSync.end())
+    {
+        KeyOpFieldsValuesTuple t = it->second;
+        string qos_name = kfvKey(t);
+        string op = kfvOp(t);
+
+        if (op == SET_COMMAND)
+        {
+            QosEntry entry;
+
+            for (auto i : kfvFieldsValues(t))
+            {
+                if (fvField(i) == "qos_id")
+                    entry.qos_id = fvValue(i);
+                else if (fvField(i) == "bw")
+                    entry.bw = to_uint<uint32_t>(fvValue(i));
+                else if (fvField(i) == "cps")
+                    entry.cps = to_uint<uint32_t>(fvValue(i));
+                else if (fvField(i) == "flows")
+                    entry.flows = to_uint<uint32_t>(fvValue(i));
+            }
+            if (addQosEntry(qos_name, entry))
+            {
+                it = consumer.m_toSync.erase(it);
+            }
+            else
+            {
+                it++;
+            }
+        }
+        else if (op == DEL_COMMAND)
+        {
+            if (removeQosEntry(qos_name))
+            {
+                it = consumer.m_toSync.erase(it);
+            }
+            else
+            {
+                it++;
+            }
+        }
+        else
+        {
+            SWSS_LOG_ERROR("Unknown operation %s", op.c_str());
+            it = consumer.m_toSync.erase(it);
+        }
+    }
+}
+
+bool DashRouteOrch::addOutboundRouting(string key, OutboundRoutingBulkContext& ctxt)
+{
+    SWSS_LOG_ENTER();
+
+    bool exists = (routing_entries_.find(key) != routing_entries_.end());
+    if (exists)
+    {
+        SWSS_LOG_ERROR("Outbound routing entry already exists for %s", key.c_str());
+        return true;
+    }
+
+    sai_outbound_routing_entry_t outbound_routing_entry;
+    outbound_routing_entry.switch_id = gSwitchId;
+    outbound_routing_entry.eni_id = eni_entries_[ctxt.eni].eni_entry.eni_id;
+    swss::copy(outbound_routing_entry.destination, ctxt.destination);
+    sai_attribute_t outbound_routing_attr;
+    vector<sai_attribute_t> outbound_routing_attrs;
+    auto& object_statuses = ctxt.object_statuses;
+
+    outbound_routing_attr.id = SAI_OUTBOUND_ROUTING_ENTRY_ATTR_DST_VNET_ID;
+    outbound_routing_attr.value.oid = gVnetNameToId[ctxt.vnet];
+    outbound_routing_attrs.push_back(outbound_routing_attr);
+
+    outbound_routing_attr.id = SAI_OUTBOUND_ROUTING_ENTRY_ATTR_OVERLAY_IP;
```

Comment:
This attribute should be valid only for VNET_DIRECT_ROUTE

BTW, how does it even pass SAI metadata validation?

Reply:
I have addressed this.